### ISSUE TITLE
Review-2018.04.12-彭翼

### ISSUE THREAD

**Issue by E-Samuel** [DIYer22/weekly-review#105]
# 基于DUNet的isprs训练报告
---
## 一、数据读入
pytorch中，调用数据集最好的方式（官方貌似都是这么用来着）就是自己把这个数据集写成一个类，然后在train文件中直接使用定义好了的（Net、loss、dataLoader等）使训练可以自由的组合各种网络和数据集，从而达到高内聚、低耦合的效果。

然后具体写一个数据类的方式，大致如下

1. 新建一个类，继承于torch.utils.data.Dataset
2. 重写父类的三个方法（\_\_init\_\_、\_\_getitem\_\_、\_\_len\_\_)

### **init:** 
主要是对各种需要使用的参数进行初始化，在这个步骤中，最好要完成以下几点

1. 形参列表中，最好指定一下原始文件的路径，比如（image\_path，label\_path）并同时附带上初试值，这样在不同的train文件中可以动态指定不同的数据集（这个类指定的是某种格式的读法，而非定死非得是某些特定数据）
2. 得指定一下transform和loader，transform是torchvision里头的数据处理模块，这里接受这样一个参数的目的和第一点一样，就是为了可以在train文件里动态指定，而loader是关于数据如何读，比如Image，是用Pillow还是opencv那些，要不要在输入的同时就变成numpy加些什么处理
3. 声明接下来要用到的成员变量，比如self.images和self.labels，将目录下每一个数据的路径都弄进去的list，方便后续使用的一些变量（init函数的常规操作嘛，初始化那些）

### **getitem:** 
关键模块，可以说是Dataset类的核心，在train文件中加载的DataLoader，将会直接调用这个方法，将其的返回值随着每次训练调出，**这里指定的返回值，就是你训练时直接能拿到的对象**，建议将必要的处理都在这里完成，等到写train文件的时候，每个item最好可以拿到直接就能用的东西

1. 在这里，通过传入的index参数和\_\_init\_\_里头的成员变量，得返回能直接训练的数据，比如 input = loader(self.images[index]) 
2. 通过index拿到特定的数据后，这里就是自由发挥了，对其读到的数据进行处理，之前\_\_init\_\_里指定的成员变量基本上都是用在这里的，最后return inputs, target

### **len** 
作用就是，return len(self.images)就好，返回数据集的总个数

### **关于isprs**
isprs的数据分为两类，一个是加拿大的城市，还有一个是德国的乡村，每个数据集都有大约30多张训练数据，分为RGB、IRRG、RGBIR三种格式（说来惭愧，目前我只用了RGB的数据）标注是一张完全给人看的图片，以不同的颜色分割了不同的区块，但训练的时候不能直接用……这里得进行一个预处理，先把它转成numpy，然后把那些什么(255,0,0)等三通道的颜色定义转换成标注的类别，不过最后还得整一次，能直接传入2DCrossEntropyLoss去算的，得是一个NxNx6的布尔矩阵（对应DeepUNet的最后一层输出）这个在后面会提到。
那个……numpy真好用
___
## 二、网络定义
这里使用的是学长的DeepUNet，我看他mxnet的代码里头，也把这个叫做resUNet，其实就是一个UNet前面加上了一个resNet的层（这里感觉哈，与其这样来个半吊子，干脆把UNet的下采样，都用resNet代替算了嘛，单一个小块，加预训练模型也插不上……而且除了深一点之外，貌似真没看到什么其他特别的地方，反正我之后打算试试，把那些卷积都用resNet的小block代替跑跑看，这样也方便加入官方跑好了的预训练模型）

具体实现网络的流程

1. 新建一个类，继承torch.nn
2. 重写父类的（\_\_init\_\_、forward）方法

### **init**
函数的开头，先调用一下父类的构造函数 `super(DUNet, self).\_\_init\_\_()`，然后以一个个成员变量的形式定义好每个层，几个卷积几个池化什么的，需要注意的有这些

1. 卷积不需要考虑feature map的大小，只要上一个卷积输出的通道数等于下一个卷积输入的通道数就好，比如conv1(3, **32**)，conv2(**32**, 64)，但如果后面跟全连接什么的，这里就得保证输入是上一层(featureMap.size x channel)个数对应，比如什么28x28x64那种，具体是多少根据padding和kernel_size自己算就好，然后不能直接把那个多维的东西传进去，得在forward()里 `x = x.view(x.size(0), -1)`，把他打成一维。（其实不用全连接也好嘛，最后来个干净利落的1x1卷积什么的）
2. 某些重复出现的结构，比如resNet中的block，UNet中的decoder、encoder，千万别像caffe一样全都写成一个巨大的一串（随便一个prototxt，动不动就2000多行……）相当丑而且不便于阅读，这里可以仿照torchvision官网上model写，比如他的resNet样例，就是很简洁的写好子模块，然后用那一个个nn.Sequential组合成大网络

### **forward**
除非被写成Sequential，否则那些在init中定义了的层，都只不过是一些游离的成员变量，无法完成训练过程，这里的forward函数是用来定义训练过程的，自己手动的把每个层自己的输入输出按照自己的需要连起来，比如你UNet上采样时需要用前面对应层下采样的输出结果，那就直接在forward函数中把需要给他的输入，自己挑挑扔进去，比如

`dec3 = self.dec3(torch.cat([dec4, F.upsample(enc2, dec4.size()[2:], mode='bilinear')], 1))`

像这样，第三个上采样层，输入是由第四个上采样层和第二个下采样层的输出组合后传入，可以很自由的任意组合层与层之间数据的传递流程

### **关于分割**
检测里头，我仔细看过loss源码的只有yolo，那里头是把每个bounding box（可以理解为一块特定大小的图片，每个物体的中心位于哪块box，在后续就得由这块检测出来）的结果，输出成一个向量，然后这个向量的第几个数字，分别代表了不同的定义，比如七个数字，第一个留空，第二个是置信度，第三个是物体类别，最后四个数字是物体的坐标，感觉定义上就有点玄幻的感觉，能这样就检测出来，也只能感慨反向传播力量的伟大了吧。（强行赋予那个向量各个数字意义，然后反向传播狂算，最后收敛……）

而分割的loss实现，看着就感觉莫名的舒服，有种那个，可以理解他目的的感觉，所谓分割，就是对像素的分类嘛，而像DUNet中，通过下采样和上采样之后，最后一层打出的，就是个和原图一样大小的feature map，不过冗杂了其他相邻像素的特征的，有上下文信息的，和原图一一对应的特征图，这里用每个像素点去对应原图上的每个点，然后根据类别输出的通道数，直接对每个像素都进行一次CrossEntropyLoss，每个通道就代表了每个类别的置信度，看上去就很有道理的样子（不知道我的理解有没有问题，看他代码什么的，感觉他好像就是这个意思来着）
___
## 三、训练脚本
最关键的核心类，目前我也没把握说自己已经完全弄懂了，只是姑且说一下自己已经了解了哪些吧

整个训练的流程哈，大致是分为这么几个步骤

1. 数据集加载
2. 网络初始化
3. loss：criterion、 优化器：optimizer、 各种训练参数：args 的初始化
4. `for epoch in range(0, args['epoch_num']):`开始迭代
5. train与test（train用来在训练集上更新网络参数，test用来在测试集上检验这个模型的普适性）

### **数据集加载**
首先，把之前写好的数据类import进来，定义好参数传入，然后定义一个DataLoader，就像

`train_loader = DataLoader(train_set, batch_size=args['train_batch_size'], num_worker=8, shuffle=True)`

以上四个参数分别是，数据集、batch\_size、子线程数、是否乱序，其他功能可以去查文档。这样，把如何从数据集中读取数据设定好，这个train\_loader对象就能直接用了（其他还有很多参数，不过我没用来着，看文档就好）

另外还有一点得提一下，一般来说，数据集初始化的代码里头会加入一个transforms模块，是从官方torchvision里头导入的一个工具类，可以用pytorch官方写好的一些数据增强函数，我目前只用了一个ToTensor()，其他的那些还在慢慢了解中……

### **网络初始化**
把之前写好的网络定义import进来，然后直接

`net = DUNet(num_classes=6).cuda()`

嗯……没了，把导入的东西实例化一下就好

### **loss、优化器、训练参数**
1. loss嘛，这个其实得单开一节说来着，毕竟官方给那些不是都能直接用，很多得自己重新实现，或者在他的基础上改改什么的，总而言之，就和之前那几个一样，把自己写好的，或者官方的导入，初始化一下（这里只是声明一下跑训练的流程）
2. optim，优化器，顾名思义，就是训练时采用的优化策略，比如说哈，不可能从头到尾一个学习率跑到死，所以每隔多少个epoch学习率下降，或者加个momentum，动量那个什么的，防止陷入局部最优（231n里讲到过）torch.optim就是这样一个优化器的集合，在训练前记得定义好
3. args，参数那些，一般都是写一大串参数名放在开头，不过之前在github上看到了一种很棒的写法，就是声明一个很大的字典，名为args，然后每个成员名对应需要的参数，比如
`args = {'train_batch_size': 16, 'lr': 1e-10}`
看上去美观了许多，而且在使用的时候，也是args['lr']那种，优雅~~

### **epoch**
>`for epoch in range(0, args['epoch_num']):`
>
>`train()`
>
>`test()`

差不多这样，一个epoch里头train和test交替着跑，不要把train()和test()里头的代码写到循环里去，会显得很乱，单独提出来整，会比较整洁的样子

### **train、test**
基本流程也是套路，就按照：

`for input, target in dataset:`//拿到一组input和target

`optimizer.zero_grad()`求解器初始化

`output = net(input)`把数据放到网络中训练，输出结果

`loss = loss_function(output, target)`把结果和真实值对比，算出loss值

`loss.backward()`反向传播

`optimizer.step()`求解器更新

差不多这些，基本就可以跑了，记得一般来说，train中改权值，test中存模型

`torch.save(net.state_dict(), os.path.join(ckpt_path, exp_name, snapshot_name + '.pth'))`
`torch.save(optimizer.state_dict(), os.path.join(ckpt_path, exp_name, snapshot_name + '.pth'))`

### **总结**
先写子模块，比如data、net、loss那些，然后再写一个train文件，把那些东西都读进来，一组组epoch更新，保存就好……差不多就这样吧
